File: src/workflows/failure_analysis.ts
import { FailureItem } from '../types/failure_item';
import { TriageResult, ROIAnalysis } from '../types/analysis';

/**
 * Main workflow for failure analysis and ROI calculation
 * Converts CI tool outputs to standardized format and applies LLM analysis
 */
export class FailureAnalysisWorkflow {
  private triagePrompt: string;
  private guardrailTemplate: string;

  constructor() {
    // These would be loaded from prompts/triage.md and prompts/guardrails.md
    this.triagePrompt = `You are an expert software engineer analyzing CI failures.
    
Analyze the following failure item and provide:
1. Classification of the error type
2. Root cause analysis 
3. Fix recommendation with code example

Input: {{failure_item}}

Output format: JSON with fields: classification, root_cause, fix_recommendation`;
    
    this.guardrailTemplate = `Always follow these constraints:
- Never output absolute file paths
- Always provide minimal, focused code examples
- Keep explanations concise but comprehensive
- Do not include any markdown formatting in your response`;
  }

  /**
   * Process a batch of CI failures through the analysis workflow
   */
  async processFailures(failures: FailureItem[]): Promise<{
    triageResults: TriageResult[];
    roiAnalysis: ROIAnalysis;
  }> {
    const triageResults: TriageResult[] = [];
    
    // Convert each failure to standardized format and analyze
    for (const failure of failures) {
      const triageResult = await this.analyzeFailure(failure);
      triageResults.push(triageResult);
    }
    
    // Generate ROI analysis using req_to_srs_roi.yaml workflow
    const roiAnalysis = await this.calculateROI(triageResults);
    
    return {
      triageResults,
      roiAnalysis
    };
  }

  /**
   * Analyze individual failure item using LLM
   */
  private async analyzeFailure(failure: FailureItem): Promise<TriageResult> {
    // In a real implementation, this would call an LLM API
    // For now, we'll return mock results based on failure content
    
    const classification = this.classifyErrorType(failure);
    const rootCause = this.analyzeRootCause(failure);
    const fixRecommendation = this.generateFix(failure);
    
    return {
      id: failure.id,
      classification,
      rootCause,
      fixRecommendation,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Classify error type based on failure content
   */
  private classifyErrorType(failure: FailureItem): string {
    if (failure.tool.includes('eslint')) {
      return 'linting_error';
    } else if (failure.tool.includes('pytest')) {
      return 'test_failure';
    } else if (failure.tool.includes('clippy')) {
      return 'rust_analysis';
    }
    return 'unknown';
  }

  /**
   * Analyze root cause of failure
   */
  private analyzeRootCause(failure: FailureItem): string {
    // Mock analysis based on error message patterns
    const msg = failure.message.toLowerCase();
    
    if (msg.includes('undefined')) {
      return 'Variable not declared or out of scope';
    } else if (msg.includes('assertion')) {
      return 'Test condition failed';
    } else if (msg.includes('unused')) {
      return 'Unnecessary code detected';
    }
    
    return 'Generic error pattern';
  }

  /**
   * Generate fix recommendation
   */
  private generateFix(failure: FailureItem): string {
    // Mock fix generation based on failure type
    const msg = failure.message.toLowerCase();
    
    if (msg.includes('undefined')) {
      return 'Declare the variable before use or check for null/undefined values';
    } else if (msg.includes('assertion')) {
      return 'Review test conditions and ensure expected values match actual results';
    } else if (msg.includes('unused')) {
      return 'Remove unused variables or add proper usage to prevent warnings';
    }
    
    return 'Review code for potential logical or syntax errors';
  }

  /**
   * Calculate ROI for each failure analysis
   */
  private async calculateROI(triageResults: TriageResult[]): Promise<ROIAnalysis> {
    // Mock ROI calculation based on severity and impact factors
    const totalCost = triageResults.reduce((sum, result) => {
      return sum + (result.classification === 'test_failure' ? 100 : 50);
    }, 0);
    
    const estimatedSavings = triageResults.reduce((sum, result) => {
      return sum + (result.classification === 'test_failure' ? 80 : 30);
    }, 0);
    
    return {
      totalCost,
      estimatedSavings,
      roi: totalCost > 0 ? (estimatedSavings - totalCost) / totalCost : 0,
      analysis: triageResults.map(result => ({
        failureId: result.id,
        cost: result.classification === 'test_failure' ? 100 : 50,
        savings: result.classification === 'test_failure' ? 80 : 30
      }))
    };
  }
}

// Export the main processing function
export async function processFailureAnalysis(
  failures: FailureItem[]
): Promise<{
  triageResults: TriageResult[];
  roiAnalysis: ROIAnalysis;
}> {
  const workflow = new FailureAnalysisWorkflow();
  return workflow.processFailures(failures);
}

File: src/workflows/failure_analysis.py
# Implementation Plan for CI Result Normalization & Workflow
#
# Phase 2 – Prompt Integration
#
# Goal:
# Integrate triage prompt and guardrails into LLM inference workflow to generate error classification and fix suggestions.
#
# Implementation:
#
# 1. Create a class `TriageAgent` that encapsulates the logic for prompting LLMs
# 2. Define input/output structures that align with `failure_item.schema.json` and `output.schema.json`
# 3. Integrate guardrail constraints into prompt template
#
# Let's create the main workflow module:
#
# ```python
# src/workflows/failure_analysis.py
from typing import List, Dict, Any
from src.schemas.failure_item import FailureItem
from src.prompts.triage import TRIAGE_PROMPT_TEMPLATE
from src.prompts.guardrails import GUARDRAILS_CONSTRAINTS

class TriageAgent:
    """LLM-based triage agent for analyzing CI failures and generating fix suggestions"""
    
    def __init__(self, model_name: str = "gpt-4"):
        self.model_name = model_name
        
    def analyze_failure(self, failure_item: FailureItem) -> Dict[str, Any]:
        """
        Analyze a single failure item and generate triage output
        """
        # Prepare input context for LLM
        context = {
            "failure": {
                "tool": failure_item.tool,
                "message": failure_item.message,
                "path": failure_item.path,
                "details": failure_item.details,
                "severity": failure_item.severity
            },
            "guardrails": GUARDRAILS_CONSTRAINTS
        }
        
        # Construct prompt
        prompt = TRIAGE_PROMPT_TEMPLATE.format(**context)
        
        # Simulate LLM response (in real implementation, this would call an LLM API)
        response = self._call_llm(prompt)
        
        return response
    
    def _call_llm(self, prompt: str) -> Dict[str, Any]:
        """
        Placeholder for actual LLM call
        In practice, this would integrate with OpenAI or other LLM APIs
        """
        # This is a mock implementation - in reality, we'd use an API client
        # For now, we'll return a dummy response structure that matches expected schema
        
        return {
            "hypothesis": f"Based on {prompt[:50]}..., the failure likely stems from a configuration issue.",
            "suspects": [
                {
                    "file": "src/main.py",
                    "line": 42,
                    "reason": "Misconfigured import path"
                }
            ],
            "patch": {
                "unified_diff": """--- a/src/main.py\n+++ b/src/main.py\n@@ -1,5 +1,5 @@\n def main():\n-    import utils\n+    from . import utils\n     return utils.process()\n""",
                "files_changed": 1,
                "lines_added": 1,
                "lines_removed": 1
            },
            "tests": [
                {
                    "path": "test_main.py",
                    "content": "def test_main():\n    assert main() == 'processed'",
                    "purpose": "Verify the fix resolves the import error"
                }
            ]
        }

# Example usage function for batch processing failures
def process_failures(failure_items: List[FailureItem]) -> List[Dict[str, Any]]:
    """
    Process multiple failure items through triage workflow
    """
    agent = TriageAgent()
    results = []
    
    for item in failure_items:
        result = agent.analyze_failure(item)
        results.append(result)
        
    return results